engine/src/main/java/org/hawkular/alerts/engine/impl/ispn/IspnAlertsServiceImpl.java

```diff
@@ -488,17 +543,110 @@ public Alert getAlert(String tenantId, String alertId, boolean thin) throws Exce
     }
 
     @Override
     public Page<Event> getEvents(String tenantId, EventsCriteria criteria, Pager pager) throws Exception {
-        return null;
+        return getEvents(Collections.singleton(tenantId), criteria, pager);
     }
 
     @Override
     public Page<Event> getEvents(Set<String> tenantIds, EventsCriteria criteria, Pager pager) throws Exception {
-        return null;
+        if (isEmpty(tenantIds)) {
+            throw new IllegalArgumentException("TenantIds must be not null");
+        }
+        boolean filter = (null != criteria && criteria.hasCriteria());
+        if (filter) {
+            log.debugf("getEvents criteria: %s", criteria.toString());
+        }
+
+        StringBuilder query = new StringBuilder("from org.hawkular.alerts.engine.impl.ispn.model.IspnEvent where ");
+        query.append("(");
+        Iterator<String> iter = tenantIds.iterator();
+        while (iter.hasNext()) {
+            String tenantId = iter.next();
+            query.append("tenantId = '").append(tenantId).append("' ");
+            if (iter.hasNext()) {
+                query.append("or ");
+            }
+        }
+        query.append(") ");
+
+        if (filter) {
+            if (criteria.hasEventIdCriteria()) {
+                query.append("and (");
+                iter = criteria.getEventIds().iterator();
+                while (iter.hasNext()) {
+                    String eventId = iter.next();
+                    query.append("id = '").append(eventId).append("' ");
+                    if (iter.hasNext()) {
+                        query.append("or ");
+                    }
+                }
+                query.append(") ");
+            }
+            if (criteria.hasTagQueryCriteria()) {
+                query.append("and (tags : ");
+                parseTagQuery(criteria.getTagQuery(), query);
+                query.append(") ");
+            }
+            if (criteria.hasTriggerIdCriteria()) {
+                query.append("and (");
+                iter = extractTriggerIds(criteria).iterator();
+                while (iter.hasNext()) {
+                    String triggerId = iter.next();
+                    query.append("triggerId = '").append(triggerId).append("' ");
+                    if (iter.hasNext()) {
+                        query.append("or ");
+                    }
+                }
+                query.append(") ");
+            }
+            if (criteria.hasCTimeCriteria()) {
+                query.append("and (");
+                if (criteria.getStartTime() != null) {
+                    query.append("ctime >= ").append(criteria.getStartTime()).append(" ");
+                }
+                if (criteria.getEndTime() != null) {
+                    if (criteria.getStartTime() != null) {
+                        query.append("and ");
+                    }
+                    query.append("ctime <= ").append(criteria.getEndTime()).append(" ");
```

Comment:
It may be useful at some point to test whether the order of operators has any perf impact.  For example, it may make sense to apply time constraints before anything else.  Hopefully there is a decent optimizer and ordering is unnecessary, but it may be worth checking.